Provide an auto wrap policy for common transformer models

zhaojuanmao opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #76455

Provide an auto wrap policy for common transformer models

Differential Revision: [D35972488](https://our.internmc.facebook.com/intern/diff/D35972488/)

zhaojuanmao committed:
Provide an auto wrap policy for common transformer models

Provide an auto wrap policy for common transformer models

Differential Revision: [D35972488](https://our.internmc.facebook.com/intern/diff/D35972488/)

[ghstack-poisoned]

rohan-varma commented on test/distributed/fsdp/test_wrap.py:
```diff
@@ -276,6 +279,25 @@ def test_always_wrap(self):
         TestFSDPWrap.NestedSequentialModel.verify_model_all_wrapped(self, model)
 
+    def test_transformer_auto_wrap_policy(self):
+        """
```
Don't need this doc if there's nothing there?

rohan-varma commented on test/distributed/fsdp/test_wrap.py:
```diff
@@ -276,6 +279,25 @@ def test_always_wrap(self):
+            auto_wrap_policy=my_auto_wrap_policy
+        )
+        self.assertTrue(isinstance(fsdp_model, FSDP))
+        for layer in fsdp_model.module.module.transformer.encoder.layers:
```
Furthermore they should be wrapped in _individual_ FSDP units right? Can we ensure that they are in distinct FSDP units?

zhaojuanmao replied:
yes, this for loop means they are individual FSDP units, otherwise layer will be the instance of original module name

rohan-varma reviewed (approved):
LGTM, thank you for the fix!

zhaojuanmao committed:
Update on "Provide an auto wrap policy for common transformer models"


Provide an auto wrap policy for common transformer models

Differential Revision: [D35972488](https://our.internmc.facebook.com/intern/diff/D35972488/)

[ghstack-poisoned]

zhaojuanmao committed:
Update on "Provide an auto wrap policy for common transformer models"


Provide an auto wrap policy for common transformer models

Differential Revision: [D35972488](https://our.internmc.facebook.com/intern/diff/D35972488/)

[ghstack-poisoned]

zhaojuanmao commented:
@pytorchbot merge this please
